include/deal.II/grid/grid_generator.h

```diff
@@ -2360,6 +2364,143 @@ namespace GridGenerator
+                 const std::vector<Point<2>> &points,
+                 const double                 iso_level,
+                 std::vector<Point<2>> &      vertices,
+                 std::vector<CellData<1>> &   cells) const;
+
+    /**
+     * Process a cell (3D).
+     */
+    void
+    process_cell(std::vector<value_type> &    ls_values,
+                 const std::vector<Point<3>> &points,
+                 const double                 iso_level,
+                 std::vector<Point<3>> &      vertices,
+                 std::vector<CellData<2>> &   cells) const;
+
+    /**
+     * Process the cell @p cell. Calls the above functions but allows the users
+     * to perform their own cell loop and potentially perform some pre- and
+     * post-processing steps.
+     */
+    void
+    process_cell(const typename DoFHandler<dim>::active_cell_iterator &cell,
+                 const VectorType &              ls_vector,
+                 const double                    iso_level,
+                 std::vector<Point<dim>> &       vertices,
+                 std::vector<CellData<dim - 1>> &cells) const;
+
+  private:
+    /**
+     * Internal function to create a quadrature rule with n_subdivisions+1
+     * equally-positioned quadrature points.
+     */
+    static Quadrature<dim>
+    create_qudrature_rule(const unsigned int n_subdivisions);
+
+    /**
+     * Process a sub-cell (2D).
+     *
+     * @note Subcells with saddle points are ignored. Please increase the number
+     *   of subdivisions in this case.
+     */
+    static void
+    process_sub_cell(const std::vector<value_type> & ls_values,
+                     const std::vector<Point<2>> &   points,
+                     const std::vector<unsigned int> mask,
+                     const double                    iso_level,
+                     std::vector<Point<2>> &         vertices,
+                     std::vector<CellData<1>> &      cells);
+
+    /**
+     * Process a sub-cell (3D).
+     */
+    static void
+    process_sub_cell(const std::vector<value_type> & ls_values,
+                     const std::vector<Point<3>> &   points,
+                     const std::vector<unsigned int> mask,
+                     const double                    iso_level,
+                     std::vector<Point<3>> &         vertices,
+                     std::vector<CellData<2>> &      cells);
+
+    /**
+     * Number of subdivisions defined in the constructor.
+     */
+    const unsigned int n_subdivisions;
+
+    /**
+     * FEValues used internally and set up with a quadrature rule with the
+     * correct number of subdivisions.
+     */
+    mutable FEValues<dim> fe_values;
+  };
+
+
+  /**
+   * Create a codim-1 triangulation based on a iso-line/surface and the
+   * marching-cube algorithm (MarchingCubeAlgorithm).
+   */
+  template <int dim, typename VectorType>
+  void
+  create_triangulation_with_marching_cube_algorithm(
+    const MarchingCubeAlgorithm<dim, VectorType> &mc,
```

Comment:
I have not actually looked at the implementation, but am unsure about this choice of interface. The `MCA` class has no virtual functions, so I assume that the first argument given here is simply of type `MCA` and not derived. If that were so, why would one require that the user creates such an object, rather than letting the user pass the necessary arguments to this function and create the object internally? If that were so, that whole class could be moved to the `.cc` file.

Reply:
> If that were so, why would one require that the user creates such an object, rather than letting the user pass the necessary arguments to this function and create the object internally?

I guess I can do that.

> If that were so, that whole class could be moved to the .cc file.

The reason for having the class being accessible by the user is that you might not want to construct a global surface grid but only a surface grid in a cell, do some postprocessing, and throw the grid away.

Personally, I am using the class this way to compute surface tensions in the case of sharp-interface methods in [adaflo](https://github.com/kronbichler/adaflo) (see: https://github.com/kronbichler/adaflo/blob/7b60179059f0cfd3ec30e7960b563595e9b0b186/include/adaflo/sharp_interface_util.h#L968-L1122):

<p align="center">
<img src="https://render.githubusercontent.com/render/math?math=(\vec{v}_i, \sigma\,\kappa\,\vec{n})_\Gamma=\sum(\vec{v}_i, \sigma\,\kappa(p_q)\,\vec{n}(p_q)\, (JxW)_q)" width="380"> 
</p>

 and @mschreter is using it to perform some surface integrals in the context of evaporation in [MeltPoolDG](https://github.com/MeltPoolDG/MeltPoolDG) (see: https://github.com/MeltPoolDG/MeltPoolDG/blob/ba9544808e2ad495424efe1317a2649bcea221aa/include/meltpooldg/evaporation/evaporation_operation_marching_cube.hpp#L153-L291).

Reply:
I don't understand how the code snippets support the current interface of `create_triangulation_with_marching_cube_algorithm` It seems to me that either I want to do something special and I need to use `MCA` but I don't need `create_triangulation_with_marching_cube_algorithm` or I want something more generic and I don't need to know that `MCA` exists